experimental/devices/mfrc522/rf522.go

```diff
@@ -0,0 +1,763 @@
+// Copyright 2016 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// Package mfrc522 controls a Mifare RFID card reader.
+// The code is largely ported from Python library : https://github.com/mxgxw/MFRC522-python
+// The datasheet is available at : https://www.nxp.com/docs/en/data-sheet/MFRC522.pdf
+package mfrc522
+
+import (
+	"errors"
+	"fmt"
+	"periph.io/x/periph/conn/gpio"
+	"periph.io/x/periph/conn/gpio/gpioreg"
+	"periph.io/x/periph/conn/spi"
+	"periph.io/x/periph/experimental/devices/mfrc522/commands"
+	"time"
+)
+
+// Dev defines the RFID reader properties, the SPI bus and GPIO pins.
+type Dev struct {
+	ResetPin      gpio.PinIO
+	IrqPin        gpio.PinIO
+	Authenticated bool
+	antennaGain   int
+	MaxSpeedHz    int
+	spiDev        spi.Conn
+}
+
+// DefaultKey provides the default bytes for card authentication for method B.
+var DefaultKey = []byte{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}
+
+// NewSPI creates and initializes the RFID card reader attached to SPI.
+// - spiPort - the SPI device to use.
+// - maxSpeed - maximum speed for SPI data transfers.
+// - resetPin - reset GPIO pin.
+// - irqPin - irq GPIO pin.
+func NewSPI(spiPort spi.Port, maxSpeed int, resetPin, irqPin string) (device *Dev, err error) {
+
+	spiDev, err := spiPort.Connect(int64(maxSpeed), spi.Mode0, 8)
+	if err != nil {
+		return
+	}
+
+	if err != nil {
+		return
+	}
+
+	dev := &Dev{
+		spiDev:     spiDev,
+		MaxSpeedHz: maxSpeed,
+	}
+
+	pin := gpioreg.ByName(resetPin)
+	dev.ResetPin = pin
+	dev.ResetPin.Out(gpio.High)
+
+	pin = gpioreg.ByName(irqPin)
+	dev.IrqPin = pin
+	dev.IrqPin.In(gpio.PullUp, gpio.FallingEdge)
+
+	err = dev.Init()
+
+	device = dev
+
+	return
+}
+
+// Init initializes the RFID chip.
+func (r *Dev) Init() (err error) {
+	err = r.Reset()
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TModeReg, 0x8D)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TPrescalerReg, 0x3E)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TReloadRegL, 30)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TReloadRegH, 0)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TxAutoReg, 0x40)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.ModeReg, 0x3D)
+	if err != nil {
+		return
+	}
+	err = r.SetAntenna(true)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (r *Dev) writeSpiData(dataIn []byte) (out []byte, err error) {
+	out = make([]byte, len(dataIn))
+	err = r.spiDev.Tx(dataIn, out)
+	return
+}
+
+func printBytes(data []byte) (res string) {
+	res = "["
+	for _, v := range data[0 : len(data)-1] {
+		res = res + fmt.Sprintf("%02x, ", byte(v))
+	}
+	res = res + fmt.Sprintf("%02x", data[len(data)-1])
+	res = res + "]"
+	return
+}
+
+func (r *Dev) devWrite(address int, data byte) (err error) {
+	newData := [2]byte{(byte(address) << 1) & 0x7E, data}
+	_, err = r.writeSpiData(newData[:])
```

Comment:
Here you can call Tx() with out being nil

Reply:
But from the docs:
```
	// For full duplex protocols (generally SPI, UART), the two buffers must have
	// the same length as both reading and writing happen simultaneously.
```

and the card will return the command/result as the first byte.

Reply:
I agree the documentation is confusing. You can use nil on one of the two if you don't care about the data (or want to send zeros).